ethereum/contracts/BeefyClient.sol

```diff
@@ -0,0 +1,474 @@
+            ),
+            "Sender must be in validator set at correct position"
+        );
+
+        // Check if validatorSignature is correct, ie. check if it matches
+        // the signature of senderPublicKey on the commitmentHash
+        require(
+            ECDSA.recover(commitmentHash, validatorSignature) == validatorPublicKey,
+            "Invalid Signature"
+        );
+
+        // Check that the bitfield actually contains enough claims to be successful, ie, >= 2/3
+        require(
+            validatorClaimsBitfield.countSetBits() >= requiredNumberOfSignatures(vset),
+            "Not enough claims in bitfield"
+        );
+
+        // Accept and save the commitment
+        validationData[nextID] = ValidationData(
+            msg.sender,
+            commitmentHash,
+            vset.id,
+            validatorClaimsBitfield,
+            block.number
+        );
+
+        emit CommitmentVerified(nextID, Phase.Initial, commitmentHash, msg.sender);
+
+        nextID = nextID + 1;
+    }
+
+    function createRandomBitfield(uint256 id) public view returns (uint256[] memory) {
+        ValidationData storage data = validationData[id];
+
+        ValidatorSet memory vset = getValidatorSet(data.validatorSetID);
+
+        // verify that block wait period has passed
+        require(block.number >= data.blockNumber + BLOCK_WAIT_PERIOD, "Block wait period not over");
+
+        return
+            Bitfield.randomNBitsWithPriorCheck(
+                getSeed(data),
+                data.validatorClaimsBitfield,
+                requiredNumberOfSignatures(vset),
+                vset.length
+            );
+    }
+
+    function createInitialBitfield(uint256[] calldata bitsToSet, uint256 length)
+        public
+        pure
+        returns (uint256[] memory)
+    {
+        return Bitfield.createBitfield(bitsToSet, length);
+    }
+
+    /**
+     * @notice Performs the final step in the validation, and then applies the commitment and optional authority update
+     * @param id an identifying value generated in the previous transaction
+     * @param commitment contains the full commitment that was used for the commitmentHash
+     * @param proof a struct containing the data needed to verify all validator signatures
+     */
+    function completeSignatureCommitment(
+        uint256 id,
+        Commitment calldata commitment,
+        ValidatorProof calldata proof
+    ) public {
+        ValidationData storage data = validationData[id];
+
+        bytes32 commitmentHash = verifyCommitment(data, commitment, proof);
+
+        latestMMRRoot = commitment.payload.mmrRootHash;
+        latestBeefyBlock = commitment.blockNumber;
+        emit NewMMRRoot(commitment.payload.mmrRootHash, commitment.blockNumber);
+
+        // Check if commitment signals an authority handover (new validator session)
+        if (commitment.validatorSetId == currentValidatorSet.id + 1) {
+            currentValidatorSet = nextValidatorSet;
```

Comment:
the biggest change in the PR is of course this, ie, we no longer seem to need to check the MMR leaves to know what the next validator set ID/len/root should be, as this data is in the main commitment payload rather than in the MMR.

is this correct?

Reply:
oh wait i see this has been moved to a separate function below?

Reply:
will need to think about this flow further - especially in a scenario where there could be a malicious node running the initial commitment to overwrite validationData[id] at any time. maybe can discuss on a quick call

Reply:
i suppose overall it may be a bit confusing with there being both nextID for validationData id as well as validatorset.id

i presume the former is used by the contracts only and the latter references the actual real validator set id?

Reply:
and so overall the flow looks like basically:

 - once complete is done, if there is a new validator set:
    - current becomes next, so next and current are both the same
    - it is now possible to only provide commitments for the new validatorset that do not have a handover
    - before any handover commitment, someone will need to call the new updateValidatorSet function to bring in the new next validator set
    - once that happens, new handover commitments are possible again and complete can be called for handover commitments again

Reply:
this all seems fine and works, but thinking about if there's a way to make it less confusing.

one idea: maybe if we had separate functions like:
 - completeSignatureCommitment -> only used for commitments in current session, never does handover, reverts if  validatorSet is not in same session
 - goToNextSession -> only used for commitments in next session, always does handover in same tx

Reply:
or maybe just use function overloading on completeSignatureCommitment, so each of the above 2 example seperate functions still have the same name, just different overload/implementation

Reply:
the above suggestion i made to make the uint64 validatorSetID on the newSignatureCommitment function a boolean may also help simplify the confusion with the different ids?

Reply:
> this all seems fine and works, but thinking about if there's a way to make it less confusing. not sure if i like the idea that the light client can exist in this "intermediary state" where currentValidatorSet === nextValidatorSet. may be better to ensure that all handovers happen atomically, without any intermediary state possible.

yeah, I agree. The handover needs some work :+1:

Reply:
obsolete